### ISSUE TITLE
LXC containers?

### ISSUE THREAD

**Issue by kzahel** [dnschneid/crouton#364]
I am wondering if it is possible to run LXC containers in my chroot. Has anyone tried this? I tried to create one in ubuntu precise but got an error "wrong fs type on /dev/shm"

I think running a container inside a chroot may not be a supported use case for lxc yet. But for chromeos it's sort of our only option, aside from doing a dev_install and some portage overlay magic?

I'm just kind of rambling here, but I couldn't find any information about LXC on chromeOS and this seemed like a reasonable enough place to find some.

**Comment by Watney** [comment]
I'm stumbling along, following after @c00w and @fwip to try out [Rocket](https://github.com/coreos/rocket). 

I successfully installed and ran ``rkt`` in /usr/local/rocket, the same directory that crouton uses. 

Next, I followed the guide to install an ACI beginning by importing the keys, but straightaway, terminal reports: ``Error adding keys: Error adding key: mkdir /etc/rkt: read-only file system`` 

How do I enable ``rkt`` to write to the directories in Chrome OS, given that (probably) all of those it's designed to use are read only?

**Comment by c00w** [comment]
When I did it there was an option to specify the directory to root things
in and I set that to my home directory. If you can't figure it out I'll do
it at home. My biggest problem was that there is too much pass through
required to use it, and it makes chromeos break. The option of less
resistance was to go fix the stuff broken in chromeos that stops me from
installing native Linux, which I still haven't done.

**Comment by Watney** [comment]
@c00w Here’s how I installed and ran Rocket
Create a folder in /usr/local/ for rocket, the same directory that crouton uses 
``sudo mkdir /usr/local/rocket`` 
Download the current tar.gz into the new folder 
``sudo wget https://github.com/coreos/rocket/releases/download/v0.3.1/rocket-v0.3.1.tar.gz`` 
Unpack
``sudo tar xzvf rocket-v0.3.1.tar.gz`` 
Check permissions and correct to ``root:root``; verify that ``rkt`` is executable 
``cd`` into ``rocket-v0.3.1`` 
``sudo ./rkt help``
This verifies that ``rkt`` is executable, and it gives you the switches and options
Create a folder in rock to save your ACIs 
``sudo mkdir /usr/local/rocket/aci`` 

**Now, for the blastoff!**  
In the same folder as ``rkt``, run this 
``sudo ./rkt --dir=/usr/local/rocket/aci --debug=true --insecure-skip-verify=true run https://github.com/coreos/etcd/releases/download/v2.0.0/etcd-v2.0.0-linux-amd64.aci ``  
To kill the Container press Ctrl ] three times

**Comment by comjf** [comment]
I've read though this entire issue, and I really appreciate you're guys commitment to trying to get this to work. My interest is with docker specifically so I can test docker deployments locally on my pixel before paying AWS costs.

It looks like @ccaapton @drinkcat @air and @edge226 spent a considerable amount of time debugging this and trying to find the solution many months ago, but could someone please summarize where we are and if this is still ongoing? hopeful? The conclusion I saw was that the required kernel flags simply aren't compiled into ChromeOS itself?

**Comment by Watney** [comment]
@comjf You might see Ron Minnich's [guide](https://docs.google.com/a/loosley.com/document/d/1WQnfloMbcR798Lk-iYPbvZt-wplW83IYyCf7SYyXTvs/edit) to install Docker. Brace yourself, Ron's bandwidth is off the charts

**Comment by mathuin** [comment]
I have an Acer Chromebook 13 (nyan-big) running arch in a chroot and I have successfully rebuilt the kernel with additional features as indicated by Docker's check-config.sh script.  I'm still not getting much success with Docker, though.

Here's the output of the script:
```
[37minfo: reading kernel config from /proc/config.gz ...[m

Generally Necessary:
- [37mcgroup hierarchy[m: [32mproperly mounted[m [/sys/fs/cgroup]
- [37mCONFIG_NAMESPACES[m: [32menabled[m
- [37mCONFIG_NET_NS[m: [32menabled[m
- [37mCONFIG_PID_NS[m: [32menabled[m
- [37mCONFIG_IPC_NS[m: [32menabled[m
- [37mCONFIG_UTS_NS[m: [32menabled[m
- [37mCONFIG_DEVPTS_MULTIPLE_INSTANCES[m: [32menabled[m
- [37mCONFIG_CGROUPS[m: [32menabled[m
- [37mCONFIG_CGROUP_CPUACCT[m: [32menabled[m
- [37mCONFIG_CGROUP_DEVICE[m: [32menabled[m
- [37mCONFIG_CGROUP_FREEZER[m: [32menabled[m
- [37mCONFIG_CGROUP_SCHED[m: [32menabled[m
- [37mCONFIG_CPUSETS[m: [32menabled[m
- [37mCONFIG_MACVLAN[m: [32menabled[m
- [37mCONFIG_VETH[m: [32menabled[m
- [37mCONFIG_BRIDGE[m: [32menabled[m
- [37mCONFIG_NF_NAT_IPV4[m: [32menabled[m
- [37mCONFIG_IP_NF_FILTER[m: [32menabled[m
- [37mCONFIG_IP_NF_TARGET_MASQUERADE[m: [32menabled[m
- [37mCONFIG_NETFILTER_XT_MATCH_ADDRTYPE[m: [32menabled[m
- [37mCONFIG_NETFILTER_XT_MATCH_CONNTRACK[m: [32menabled[m
- [37mCONFIG_NF_NAT[m: [32menabled[m
- [37mCONFIG_NF_NAT_NEEDED[m: [32menabled[m
- [37mCONFIG_POSIX_MQUEUE[m: [32menabled[m

Optional Features:
- [1mCONFIG_MEMCG_SWAP[m: [1;31mmissing[m
- [1mCONFIG_RESOURCE_COUNTERS[m: [1;31mmissing[m
- [37mCONFIG_CGROUP_PERF[m: [32menabled[m
- Storage Drivers:
  - "[34maufs[m":
    - [1mCONFIG_AUFS_FS[m: [1;31mmissing[m
    - [37mCONFIG_EXT4_FS_POSIX_ACL[m: [32menabled[m
    - [37mCONFIG_EXT4_FS_SECURITY[m: [32menabled[m
  - "[34mbtrfs[m":
    - [37mCONFIG_BTRFS_FS[m: [32menabled[m
  - "[34mdevicemapper[m":
    - [37mCONFIG_BLK_DEV_DM[m: [32menabled[m
    - [37mCONFIG_DM_THIN_PROVISIONING[m: [32menabled[m
    - [37mCONFIG_EXT4_FS[m: [32menabled[m
    - [37mCONFIG_EXT4_FS_POSIX_ACL[m: [32menabled[m
    - [37mCONFIG_EXT4_FS_SECURITY[m: [32menabled[m
  - "[34moverlay[m":
    - [1mCONFIG_OVERLAY_FS[m: [1;31mmissing[m
    - [37mCONFIG_EXT4_FS_SECURITY[m: [32menabled[m
    - [37mCONFIG_EXT4_FS_POSIX_ACL[m: [32menabled[m
```

As you can see, the "generally necessary" bits are all enabled as are the btrfs and devicemapper bits.  In theory, this would mean I could use either option as a storage driver.  I also have the lxc package installed, so I could use either lxc or native exec drivers.  

When I run the docker daemon with all the defaults (`docker -d`), I am able to pull containers like `busybox`.  However, I can't run them.  Here is the output when executing the run command:

```
[jmt@localhost ~]$ sudo docker run --rm -it busybox /bin/bash
FATA[0000] Error response from daemon: Cannot start container e68722942d423dc0f26dbe848211554c2f2bcc0a7e093222905759d281e0b34d: mountpoint for cpuset not found
[jmt@localhost ~]$
```

And here is the output from the daemon:

```
[jmt@localhost ~]$ sudo docker -d
INFO[0000] +job serveapi(unix:///var/run/docker.sock)
INFO[0000] Listening for HTTP on unix (/var/run/docker.sock)
INFO[0000] +job init_networkdriver()
INFO[0000] -job init_networkdriver() = OK (0)
INFO[0000] WARNING: mountpoint for memory not found

INFO[0000] Loading containers: start.
..........
INFO[0000] Loading containers: done.
INFO[0000] docker daemon: 1.5.0 a8a31ef; execdriver: native-0.2; graphdriver: devicemapper
INFO[0000] +job acceptconnections()
INFO[0000] -job acceptconnections() = OK (0)
INFO[0007] POST /v1.17/containers/create
INFO[0007] +job create()
INFO[0008] +job log(create, e68722942d423dc0f26dbe848211554c2f2bcc0a7e093222905759d281e0b34d, busybox:latest)
INFO[0008] -job log(create, e68722942d423dc0f26dbe848211554c2f2bcc0a7e093222905759d281e0b34d, busybox:latest) = OK (0)
INFO[0008] -job create() = OK (0)
INFO[0008] POST /v1.17/containers/e68722942d423dc0f26dbe848211554c2f2bcc0a7e093222905759d281e0b34d/attach?stderr=1&stdin=1&stdout=1&stream=1
INFO[0008] +job container_inspect(e68722942d423dc0f26dbe848211554c2f2bcc0a7e093222905759d281e0b34d)
INFO[0008] -job container_inspect(e68722942d423dc0f26dbe848211554c2f2bcc0a7e093222905759d281e0b34d) = OK (0)
INFO[0008] +job attach(e68722942d423dc0f26dbe848211554c2f2bcc0a7e093222905759d281e0b34d)
INFO[0008] POST /v1.17/containers/e68722942d423dc0f26dbe848211554c2f2bcc0a7e093222905759d281e0b34d/start
INFO[0008] +job start(e68722942d423dc0f26dbe848211554c2f2bcc0a7e093222905759d281e0b34d)
INFO[0008] +job allocate_interface(e68722942d423dc0f26dbe848211554c2f2bcc0a7e093222905759d281e0b34d)
INFO[0008] -job allocate_interface(e68722942d423dc0f26dbe848211554c2f2bcc0a7e093222905759d281e0b34d) = OK (0)
INFO[0008] +job log(start, e68722942d423dc0f26dbe848211554c2f2bcc0a7e093222905759d281e0b34d, busybox:latest)
INFO[0008] -job log(start, e68722942d423dc0f26dbe848211554c2f2bcc0a7e093222905759d281e0b34d, busybox:latest) = OK (0)
INFO[0008] +job release_interface(e68722942d423dc0f26dbe848211554c2f2bcc0a7e093222905759d281e0b34d)
INFO[0008] -job release_interface(e68722942d423dc0f26dbe848211554c2f2bcc0a7e093222905759d281e0b34d) = OK (0)
INFO[0008] -job attach(e68722942d423dc0f26dbe848211554c2f2bcc0a7e093222905759d281e0b34d) = OK (0)
INFO[0008] +job release_interface(e68722942d423dc0f26dbe848211554c2f2bcc0a7e093222905759d281e0b34d)
INFO[0008] -job release_interface(e68722942d423dc0f26dbe848211554c2f2bcc0a7e093222905759d281e0b34d) = OK (0)
ERRO[0008] Warning: error unmounting device e68722942d423dc0f26dbe848211554c2f2bcc0a7e093222905759d281e0b34d: UnmountDevice: device not-mounted id e68722942d423dc0f26dbe848211554c2f2bcc0a7e093222905759d281e0b34d
INFO[0008] +job log(die, e68722942d423dc0f26dbe848211554c2f2bcc0a7e093222905759d281e0b34d, busybox:latest)
INFO[0008] -job log(die, e68722942d423dc0f26dbe848211554c2f2bcc0a7e093222905759d281e0b34d, busybox:latest) = OK (0)
Cannot start container e68722942d423dc0f26dbe848211554c2f2bcc0a7e093222905759d281e0b34d: mountpoint for cpuset not found
INFO[0008] -job start(e68722942d423dc0f26dbe848211554c2f2bcc0a7e093222905759d281e0b34d) = ERR (1)
ERRO[0008] Handler for POST /containers/{name:.*}/start returned error: Cannot start container e68722942d423dc0f26dbe848211554c2f2bcc0a7e093222905759d281e0b34d: mountpoint for cpuset not found
ERRO[0008] HTTP Error: statusCode=500 Cannot start container e68722942d423dc0f26dbe848211554c2f2bcc0a7e093222905759d281e0b34d: mountpoint for cpuset not found
```

A similar issue is discussed in docker/docker#6681 and one of the collaborators posts a request for certain commands to be run, to indicate which cgroup objects are available (I think).  I ran those same commands, and here is the resulting output:

```
[jmt@localhost ~]$ cat /proc/cgroups
#subsys_name    hierarchy       num_cgroups     enabled
cpuset  0       1       1
cpu     1       8       1
cpuacct 0       1       1
devices 0       1       1
freezer 2       9       1
perf_event      0       1       1
[jmt@localhost ~]$ cat /proc/self/cgroup
2:freezer:/
1:cpu:/
[jmt@localhost ~]$ grep cgroup /proc/mounts
none /sys/fs/cgroup tmpfs rw,nosuid,nodev,noexec,relatime,mode=755 0 0
cgroup /sys/fs/cgroup/cpu cgroup rw,nosuid,nodev,noexec,relatime,cpu 0 0
cgroup /sys/fs/cgroup/freezer cgroup rw,nosuid,nodev,noexec,relatime,freezer 0 0
none /sys/fs/cgroup tmpfs rw,nosuid,nodev,noexec,relatime,mode=755 0 0
cgroup /sys/fs/cgroup/cpu cgroup rw,nosuid,nodev,noexec,relatime,cpu 0 0
cgroup /sys/fs/cgroup/freezer cgroup rw,nosuid,nodev,noexec,relatime,freezer 0 0
```

The `cpuset` cgroup subsystem is enabled but for some reason is not mounted nor is it listed in `/proc/self/cgroup`.  I believe this is what Docker is looking for but cannot find.  I do not yet know how to fix this problem, but I am working on it.

**Comment by mathuin** [comment]
It works!  So far.  Under the most minimal of testing.  But it works.

I wrote the following script to properly mount all the cgroups:

````
#!/bin/bash
for subsys in `cut -f1 < /proc/cgroups | grep -v subsys_name`; do
  mnt=/sys/fs/cgroup/$subsys
  if [ ! -e $mnt ]; then
    mkdir $mnt
    mount -t cgroup none $mnt -o $subsys
  fi
done
````

I also modified the kernel to support the memory cgroup.  This required adding new config variables:

````
CONFIG_MEMCG_SWAP=y
CONFIG_MEMCG_SWAP_ENABLED=y
````

I started up my arch chroot and ran `docker -d` in one window as root, then ran the following command in another window of the same chroot:

````
docker run --rm -it armv7/armhf-ubuntu:14.10
````

I got a bash prompt from inside a container.  https://github.com/djmaze/armhf-ubuntu-docker/README.md suggests tagging releases so existing Dockerfiles referencing `FROM ubuntu` and the like work correctly.

I have *not* done any of the following yet:
* constructed an image from a Dockerfile
* linked containers together
* tried fig^Wdocker compose

But this is very exciting and I wanted to share it with you. :-)

**Comment by comjf** [comment]
@Watney Thanks for that awesome guide, however, I am skeptical of following because of the warning section. That it will break things like netflix going forward. Is there no way of just updating the kernel without switching to ChromiumOS? I want my cake and to eat it too haha.

@mathuin This is great news, but when you say things like you *modified the kernel to support* what exactly do you mean by that? Does this involve following the steps for installing ChromiumOS? Also have you thought about writing a blog post about this? I am guessing with the new chromebooks (especially pixel 2) many more people are interested in making a true developer laptop.

**Comment by mathuin** [comment]
@comjf I documented the process I used to build the kernel here:  http://www.reddit.com/r/chromeos/comments/307lpx/need_help_building_and_installing_kernel_on_acer/cpuvsky  This does not mean following the steps for installing ChromiumOS but instead something different.  I may make a blog post about what I've learned through my work blog, and if I do I will ping you here and let you know. :-)

**Comment by Watney** [comment]
@mathuin Docker ... totally amazing! Really nice work with the kernel

**Comment by bendavis78** [comment]
@Watney  @fwip

It seems that rkt can run docker images: https://github.com/coreos/rkt/blob/master/Documentation/running-docker-images.md

I tried this myself to see if I could run an archlinux image, but alas it failed: 

    $ sudo rkt --dir=/usr/local/rkt/aci --debug --insecure-skip-verify fetch docker://logankoester/archlinux
    rkt: fetching image from docker://logankoester/archlinux
    rkt: warning: signature verification has been disabled
    Downloading layer: 626e4432ce9d50a71be849f7ebc7b0df14539eb93d57e0598e6144925df2cefd
    Downloading layer: f7705805f4fbadb202715bf3c75b15e6b34382b19363586c21d06c527642f9f6
    Downloading layer: d8c5babd88e24549413cf2a862b8795de32b619a583e069c7d7c8bc849f814ae
    Downloading layer: dce0559daa1b07151a9af703077f50cbe8300bee79707fe663118c098e73269e
    Downloading layer: 9b0516337e5a10feed4b9e2d312aeec22e745da6bcc19401c9987cfaa0ab7f8e
    Downloading layer: 511136ea3c5a64f264b78b5433614aec563103b4d4702f3ba7d4d2698e22c158
    sha512-4a1ecbd991bf7c82be13309382770140

    $ sudo rkt --dir=/usr/local/rkt/aci --debug --insecure-skip-verify run sha512-4a1ecbd991bf7c82be13309382770140
    2015/04/20 14:16:52 Preparing stage1
    2015/04/20 14:16:53 Loading image sha512-4a1ecbd991bf7c82be133093827701402dc23ddab35bff5144c5889568bbd432
    run: error setting up stage0: error: image sha512-4a1ecbd991bf7c82be133093827701402dc23ddab35bff5144c5889568bbd432 has no app section

**Comment by Watney** [comment]
@bendavis78 I used ``fetch docker://pritunl/archlinux`` and rocket similarly returned ``has no app section`` Given that I don't know what I'm doing and that I don't understand much about the Arch docker image, it would be pretty much luck if I got a command line (*shrug*)

Next, I moved on to Ubuntu by substituting ``fetch docker://ubuntu`` followed by 

``sudo ./rkt --dir=/usr/local/rkt/aci --debug --insecure-skip-verify run sha512-82f05b9f14e785e54fa7a674e4e0aaaa /bin/echo 'Hello world'``

Rocket successfully launched, but abruptly shut down reporting: 

``Service exit-watcher.service is not needed anymore. Stopping.
Shutting down.``

This is apparently a [known issue](https://github.com/coreos/rkt/issues/562) with a confirmed workaround. However, I haven't been able to get it to work, which says more about me than discussion over at [rkt](https://github.com/coreos/rkt).

Please let us know if you make any progress!!

**Comment by perpen** [comment]
I installed a custom kernel on my Pixel 2015, with options similar to @mathuin 's above, ie. enough for docker to use the devicemapper storage backend.
I am able to start/run containers (well, only tested with 1 which runs a JIRA server), only the "docker exec -it <container> bash" seems to be failing so far.

To help others interested, I put the scripts I used to compile/install the kernel from my archlinux chroot here: https://gist.github.com/perpen/b07606f91a7dcbafcaf8
I voluntarily do not document the scripts as I don't want anybody to run them blindly - that would pretty much brick your chromebook. I may clean write robust/portable scripts next weekend.

@mathuin did you play some more with docker?

**Comment by mathuin** [comment]
I confess that I stopped playing with docker on the Chromebook after realizing that the kernel changes I had made were causing some stability issues.  I recommend running the docker daemon in debug mode to see why the exec command fails, that was helpful to me.

**Comment by perpen** [comment]
Thanks Jack. I have been running the custom kernel for 3 days and didn't notice stability issues - but then I didn't use docker much.
May be some other option you may have enabled - you could compare your kernel config with the one in my gist linked above. 
I run the daemon with "-D" but it is not helping much. This is the error I get when doing `docker exec -it jira6 bash`:
```
INFO[0080] -job log(exec_start: bash , a9ca5bde6de932d81b3d59abc717e6fab9a0fc4686ab525aeffc4013809f4b4e, jira6:latest) = OK (0) 
DEBU[0080] attach: stdin: begin                         
DEBU[0080] attach: stdout: begin                        
DEBU[0080] attach: stderr: begin                        
DEBU[0080] Calling POST /exec/{name:.*}/resize          
INFO[0080] POST /v1.18/exec/10e615a3664c59170cef11078235897ab06665df11f62b1327454e8cd5e4b014/resize?h=28&w=193 
INFO[0080] +job execResize(10e615a3664c59170cef11078235897ab06665df11f62b1327454e8cd5e4b014, 28, 193) 
WARN[0080] exit status 1                                
ERRO[0080] Error running command in existing container a9ca5bde6de932d81b3d59abc717e6fab9a0fc4686ab525aeffc4013809f4b4e: [8] System error: open /proc/158/status: no such file or directory 
DEBU[0080] Exec task in container a9ca5bde6de932d81b3d59abc717e6fab9a0fc4686ab525aeffc4013809f4b4e exited with code -1 
DEBU[0080] attach: stdout: end                          
DEBU[0080] attach: stderr: end                          
bad file descriptor
```
I'll try some stracing before posting to some docker group.

**Comment by gdm85** [comment]
I see the discussion has been hijacked on Docker :)
I have some experience with Docker, and although I am a big fan of it for bigger machines I don't think it should be used at all for crouton. The reason is, Docker is a container engine for **managing applications**, which is a good fit for servers. Here we have a chromebook, and we want to still use it as a minidesktop-tablet-you-name-it, so the added complexity of Docker is not worth it.

I think we should stick with LXC which is native, or maybe some container technology which is more desktop-oriented? I know [Ubuntu has been working on something like that](https://insights.ubuntu.com/2015/04/22/here-comes-kilo-15-05-containers-will-never-be-the-same-again/), although I didn't try it.

Docker is totally not a replacement for full systems chroots, and although somebody here mentioned systemd I would skip that problem completely for now (and use hacks as @dnschneid said), until cgroups support (both in-kernel and userland) is stabilized.
Once again, the right tool for the right job.

**Comment by perpen** [comment]
Maybe to you the chromebook is just for "minidesktop"/tablet stuff, but I (and many others) use it differently.

My personal use case: I have a collection of dockerfiles I use to bring up apps under different configurations (build servers, Atlassian tools, Nexus, etc), to test custom code or various integrations. Great use case for docker. 
Without docker on my Pixel I had to remote into other boxes (e.g. Digital Ocean nodes) to use my dockerfiles. Not anymore.
Plus... I *like* docker and I enjoy using it :)

Beyond this, crouton allows us to make our chromebooks into proper development environments for a variety of technologies, and I don't see why we should limit ourselves to what works right now out of the box.

**Comment by jpw1116** [comment]
Same here. I am still opting for an easy to spin up, destroyable local dev environment. The question still seems to be if a native Chrome OS solution will ever arise. Otherwise, what's the simplest chroot method?

**Comment by gdm85** [comment]
@jpw1116 the simplest chroot method is...crouton?!

@perpen by reading [this comment](https://github.com/dnschneid/crouton/issues/364#issuecomment-24436430) I thought that there was even an idea at some point to use containers **instead** of crouton's chroot, which I think it's not practical.

If you want to use it from **within** the chroot then it's your business :) inception anyone?

**Comment by bendavis78** [comment]
@gdm85, a major advantage of using LXC would be the ability to use systemd, which requires a PID namespace (right now chromeos/crouton doesn't support this out of the box). For better or worse, Linux distros are moving in this direction (including Ubuntu). More and more applications will come to rely on systemd, so it's inevitable that we'll need a chroot solution that supports PID namespacing.

**Comment by jpw1116** [comment]
@gdm85, by simplest I meant a *functioning* Docker instance via chroot. Of course, I've been running an LTS in crouton for ages.

**Comment by gdm85** [comment]
Ok, so it was not my impression, the discussion has really gone OT... :)

1) the original topic is about how to run containers (LXC) **within** the crouton chroot
2) @bendavis78 if I understood correctly you would like to replace the chroot with a container, may I suggest that goes in a separate issue?
3) @jpw1116 nowadays Docker doesn't even run LXC anymore by default, may I suggest another issue too? Because there might be more Docker-specific issues arising also over time, as both crouton and Docker are in high flux development

**Comment by perpen** [comment]
I apologise @gdm85, but I am going to confuse things even more wrt containers being inside or outside the chroot :) 

I am now running the docker daemon outside my chroot, but I use it (as normal) from inside the chroot.
I run the following script on Chrome OS boot: https://gist.github.com/perpen/980743bb39410e3ac9f6
All seems fine so far, at least for the scenarios I tried. And it fixed my problems with `docker exec`!
Of course this still requires a custom kernel :(

Ofc having the daemon in Chrome OS made me think of using docker as a replacement for the current chroots. Taking inspiration from https://github.com/maci0/docker-systemd-unpriv I tried to create a non-privileged archlinux container running systemd and I failed, probably b/c of my lack of understanding of systemd.
I started this same container just once with `--privileged`, and it seemed to work at least a bit, e.g. `journalctl` worked. But this had undesired side-effects (changed the audio volume?!, remounted a USB key) and I gave up at this point as I don't like the privileged idea. I'll explore again some other weekend.

**Comment by perpen** [comment]
Well, I just couldn't resist. I have a seemingly functioning non-privileged docker container with systemd.

Yesterday I was unable to view any systemd output, but tonight I started the container with process `strace -f systemd --system` and at least I could see what the problems were.
I had to do manually some of the things systemd couldn't do in a non-privileged container: I created in Chrome OS a dedicated `/run-systemd` tmpfs mount point, mapped to volume `/run`. And I mounted in Chrome OS `/sys/fs/cgroup/systemd` using options I got from a working archlinux vm, and mapped `/sys/fs/cgroup` into the container.
Then systemd ran normally.
Interestingly it seems some systemd units now check for required linux capabilities before starting. I suppose this is intended for use in containers, nice to see them thinking of the use case.

Was fun but I'm probably not going to be using this at all! But that made me curious about the possibility of using minijail or other compact solutions and have the option of using systemd in our chroots.
I searched the issues but I'm still unsure about how far we are with this.

**Comment by gdm85** [comment]
@perpen I feel like systemd and cgroups have still a long way to go before being mature..and you can't expect non-breaking changes in Docker either ;)

**Comment by perpen** [comment]
This pm I got an archlinux/systemd chroot running using nsinit, with mount/ipc/uts/pid namespaces. `nsinit` is a utility which comes with libcontainer, the low-level lib docker uses to create containers.
I am running a custom kernel but looking at the config for the standard Chrome OS kernel it seems to have all the basic namespacing stuff enabled - so I am not expecting problems with a vanilla crouton system.
I'm going on hols soon but after that I'll try making a PoC replacement for the current chroot mechanism.

**Comment by ccaapton** [comment]
@perpen Did you try nsinit in unprivileged mode, aka uid_map/gid_map stuff? I'm quite interested in running everything without root. I tried nsinit a couple of times with the stock kernel(C720 on 3.8.11) without luck.

Currently I'm running everything unprivileged with "unshare -Umpf" on stock kernel with some custom scripts(pivot_root etc). Most stuff works, including Xserver, but it is kind of adhoc.

**Comment by bendavis78** [comment]
@perpen, could you post the container.json you used with nsinit?

**Comment by olofj** [comment]
Hi,

Just a quick heads up here about recent changes in the Chrome OS kernel. We've enabled some of the needed options to run Docker/LXC on R45 (current ToT).

We've only enabled this on kernel versions 3.14 and 3.18. Container support isn't deemed solid in kernels before that, so unfortunately those platforms won't see it enabled.

For reference, the platforms running the 3.14 kernel are Rockchip, the 2015 Chromebook Pixel and other Broadwell-based Chromebooks.

**Comment by Watney** [comment]
*kaboom* — @olofj you're da bomb!

**Comment by mathuin** [comment]
Is there any chance that these kernel versions will trickle down to current ARM-based Chromebooks?

**Comment by olofj** [comment]
>
> Is there any chance that these kernel versions will trickle down to
> current ARM-based Chromebooks?
>
There are no plans to update to newer kernels on existing platforms. But
that's off topic for this thread.

Note that the recently introduced RK3288-based Chromebooks run 3.14.

**Comment by perpen** [comment]
@olofj I'm really excited about your news, can't wait to try it!

@ccaapton no I didn't try uid remapping, I suppose we could remap chronos to root but I'm not sure about the benefit? I would quite like being able to load modules from the container for example (although I didn't test that).
Did you get xiwi working? That's where I got stuck, as iirc the chroot-side process was trying to communicate with the chrome side via `/proc/<pid>/maps`, which obv. doesn't work with the pid namespace.

@bendavis78 for the container config I have been using you can have a look at https://github.com/perpen/chromeos-nsinit. As I said earlier I did not test nsinit with the vanilla kernel, so if you try it please let me know the result.

**Comment by tonyxue** [comment]
@olofj That's a good news. And it's nice to see you show up here to tell us that. :)

**Comment by wenlock** [comment]
wondering if anyone has written a howto for this yet?   with latest docker, here is how far i get

sudo docker -d --storage-driver=vfs --iptables=false -g ~/data/docker
WARN[0000] You are running linux kernel version 3.8.11, which might be unstable running docker. Please upgrade your kernel to 3.10.0. 
INFO[0000] Listening for HTTP on unix (/var/run/docker.sock) 
WARN[0000] Running modprobe bridge nf_nat failed with message: , error: exit status 1 
WARN[0000] Your kernel does not support cgroup memory limit: mountpoint for memory not found 
FATA[0000] Error mounting devices cgroup: mountpoint for devices not found

Any ideas on how to solve the cgroup mount error?

and with -D output

sudo docker -d --storage-driver=vfs --iptables
=false -g ~/data/docker -D
DEBU[0000] Registering GET, /containers/{name:.*}/top   
DEBU[0000] Registering GET, /containers/{name:.*}/stats 
DEBU[0000] Registering GET, /info                       
DEBU[0000] Registering GET, /version                    
DEBU[0000] Registering GET, /containers/ps              
DEBU[0000] Registering GET, /images/get                 
DEBU[0000] Registering GET, /images/{name:.*}/get       
DEBU[0000] Registering GET, /images/{name:.*}/history   
DEBU[0000] Registering GET, /images/{name:.*}/json      
DEBU[0000] Registering GET, /_ping                      
DEBU[0000] Registering GET, /events                     
DEBU[0000] Registering GET, /images/json                
DEBU[0000] Registering GET, /containers/{name:.*}/changes 
DEBU[0000] Registering GET, /containers/{name:.*}/logs  
DEBU[0000] Registering GET, /exec/{id:.*}/json          
DEBU[0000] Registering GET, /containers/{name:.*}/json  
DEBU[0000] Registering GET, /containers/{name:.*}/attach/ws 
DEBU[0000] Registering GET, /images/search              
DEBU[0000] Registering GET, /containers/json            
DEBU[0000] Registering GET, /containers/{name:.*}/export 
DEBU[0000] Registering POST, /build                     
DEBU[0000] Registering POST, /containers/{name:.*}/wait 
DEBU[0000] Registering POST, /containers/{name:.*}/attach 
DEBU[0000] Registering POST, /containers/{name:.*}/copy 
DEBU[0000] Registering POST, /containers/{name:.*}/exec 
DEBU[0000] Registering POST, /exec/{name:.*}/start      
DEBU[0000] Registering POST, /auth                      
DEBU[0000] Registering POST, /images/create             
DEBU[0000] Registering POST, /images/load               
DEBU[0000] Registering POST, /images/{name:.*}/push     
DEBU[0000] Registering POST, /images/{name:.*}/tag      
DEBU[0000] Registering POST, /containers/{name:.*}/pause 
DEBU[0000] Registering POST, /containers/{name:.*}/rename 
DEBU[0000] Registering POST, /commit                    
DEBU[0000] Registering POST, /containers/{name:.*}/restart 
DEBU[0000] Registering POST, /exec/{name:.*}/resize     
DEBU[0000] Registering POST, /containers/{name:.*}/unpause 
DEBU[0000] Registering POST, /containers/{name:.*}/kill 
DEBU[0000] Registering POST, /containers/{name:.*}/start 
DEBU[0000] Registering POST, /containers/{name:.*}/stop 
DEBU[0000] Registering POST, /containers/{name:.*}/resize 
DEBU[0000] Registering POST, /containers/create         
DEBU[0000] Registering DELETE, /containers/{name:.*}    
DEBU[0000] Registering DELETE, /images/{name:.*}        
DEBU[0000] Registering OPTIONS,                         
WARN[0000] You are running linux kernel version 3.8.11, which might be unstable running docker. Please upgrade your kernel to 3.10.0. 
DEBU[0000] [graphdriver] trying provided driver "vfs"   
DEBU[0000] Using graph driver vfs                       
DEBU[0000] Using default logging driver json-file       
DEBU[0000] Creating images graph                        
DEBU[0000] Restored 0 elements                          
DEBU[0000] Creating repository list                     
DEBU[0000] docker group found. gid: 999                 
INFO[0000] Listening for HTTP on unix (/var/run/docker.sock) 
WARN[0000] Running modprobe bridge nf_nat failed with message: , error: exit status 1 
DEBU[0000] /sbin/iptables, [--wait -t nat -D PREROUTING -m addrtype --dst-type LOCAL -j DOCKER] 
DEBU[0000] /sbin/iptables, [--wait -t nat -D OUTPUT -m addrtype --dst-type LOCAL ! --dst 127.0.0.0/8 -j DOCKER] 
DEBU[0000] /sbin/iptables, [--wait -t nat -D OUTPUT -m addrtype --dst-type LOCAL -j DOCKER] 
DEBU[0000] /sbin/iptables, [--wait -t nat -D PREROUTING] 
DEBU[0000] /sbin/iptables, [--wait -t nat -D OUTPUT]    
DEBU[0000] /sbin/iptables, [--wait -t nat -F DOCKER]    
DEBU[0000] /sbin/iptables, [--wait -t nat -X DOCKER]    
WARN[0000] Your kernel does not support cgroup memory limit: mountpoint for memory not found 
FATA[0000] Error mounting devices cgroup: mountpoint for devices not found

I've been using docker in virtualbox on my chrombook for a bit, and would love to be able to move out of virtualbox.

 sudo docker version
Client version: 1.7.0
Client API version: 1.19
Go version (client): go1.4.2
Git commit (client): 0baf609
OS/Arch (client): linux/amd64

**Comment by vevmesteren** [comment]
Given that there was no answer since June 27th I guess we shouldn't be holding our breath for LXC to come to life inside a crouton chroot? No major architectural overhaul planned by chance. Will use chroot to code but will probably need to lxc-start my containers on a seperate box...can't give up on the sheer speed this chrome os chroot combo offers...

**Comment by DennisLfromGA** [comment]
An interesting question has been posed in the 'Chromium OS Discuss forum' - [Docker support in kernel](https://groups.google.com/a/chromium.org/forum/#!topic/chromium-os-discuss/KCXncqE2SDc).
Since it may depend on kernel 3.10+, it won't work on a lot of the older Chromebooks but it'll be interesting to see what the Developers say about it.

**Comment by cat-turner** [comment]
Did anyone get this working? I have chrome book pixel, which is supposedly supported by the update. I am on the beta channel, do I have to get on the dev channel for this? I can
t seem to find a way to get it working.

**Comment by HLFH** [comment]
In fact, Chrome OS is an obsolete platform. Google gave up on developers a long time ago. Let's see what it will be with the Andromeda project. But IMHO, you will lose less time by migrating to Linux/macOS/Windows platforms.

**Comment by cat-turner** [comment]
@HLFH bummer, I really liked working in it. It's been good. I got by working on c9 and crouton crust for a while but now I need to get real work done. Welp, guess I'm going with Linux. I'm diving into Linux Mint on my old PC.

**Comment by dnschneid** [comment]
[\[citation needed\]](https://en.wikipedia.org/wiki/Wikipedia:Citation_needed)

**Comment by cat-turner** [comment]
I thought this is their way of getting developers to use their google cloud platform? I get it, but I can't pay into that. Maybe if their next chromebook/andromedia-like machine included X years of cloud support on it, but that still doesn't address the problem of dev needs being met secondarily. It's getting in the way of my work trying to find work around, but it's part of the dev life I suppose.

**Comment by HLFH** [comment]
@dnschneid Chrome Dev Editor is [discontinued](https://github.com/GoogleChrome/chromedeveditor). No [Arch Linux support for Crouton](https://github.com/dnschneid/crouton/pull/1201). And apparently, Google gave up not only on developers but also on the American people: [Eric Schmidt & Google were supporting the globalist Clinton](http://qz.com/823922/eric-schmidt-played-a-crucial-role-in-team-hillarys-election-tech/): we've seen a lot of people showing evidence that Google Search was biased in favor of Hillary Clinton during the Democratic Party presidential primaries, and during the US presidential election.

**Comment by fwip** [comment]
I think that this is not the appropriate forum for your political complaints.

**Comment by DennisLfromGA** [comment]
Chrome Dev Editor is discontinued. No Arch Linux support for Crouton. ≠ Chrome OS is an obsolete platform

**Comment by dnschneid** [comment]
I was mostly just disappointed that no political candidate made a platform of pushing a timeline for crouton to use containers, and that nobody in the media commented on the candidates' preferred proportion of salad to croutons.  Priorities, people!

**Comment by vevmesteren** [comment]
![image](https://cloud.githubusercontent.com/assets/143454/20502749/9530b66a-b00d-11e6-9ec1-436cd10295c9.png)

**Comment by vevmesteren** [comment]
The media conspirators just didn't pay attention. LOCK THEM UP!

![image](https://cloud.githubusercontent.com/assets/143454/20502771/ab2fb43e-b00d-11e6-80b2-fbcac22807b3.png)

**Comment by tedm** [comment]
I don't see the big deal about lxc containers. Few chromebooks are powerful enough to really use containers and VMs IMHO.  On an i7 desktop with 16GB of RAM, I use docker and core os, and virtualbox, but I wouldn't see a big use of these on chromebooks.

I'm waiting for my Toshiba 2 to be able to run Android apps. It's not a deal killer if it doesn't, as this doesn't have a touch screen.

I read that Trump (uggh) uses a Samsung phone, but his staffers use iphones. I am not sure what I'm going to get after my Nexus 6, but probably not the Pixel, just too pricey. I like the security of the iphones, but not a big apple fan either.

**Comment by tedm** [comment]
re: arch - I got my Pine64 1.2Ghz / 1GB computer, only like 6 months after I paid for it... anyways, I was bummed that there was little Arch support for it, as that is what I would want to. But a BSP and version called Longsheep ubuntu loaded right up, and works fine. It's at work now as a dedicated nInvaders character based game system. Not bad for $19.00.

**Comment by tonyxue** [comment]
Uh, this long and old thread pops up again.
I'd say Chrome OS isn't obsolete at all. Manufacturers are still building better and higher-end chromebooks. 

Containers probably won't be possible on chromebooks in the short run as the hardware limitations and, more importantly, if containers are directly available on chromebook one day, we probably won't need crouton any more.

**Comment by DennisLfromGA** [comment]
@tonyxue,

I don't want to add too much more to this old thread but some recent development has added some fuel to it, inadvertantly maybe.

> Containers probably won't be possible on chromebooks in the short run as the hardware limitations 

In fact ARC++ is Android apps (Google Play) in a container on CrOS and it works extremely well, my Dell Chromebook 13 uses it almost flawlessly. I say almost because 'nougat' hasn't hit yet which allows resizing windows and such.

> and, more importantly, if containers are directly available on chromebook one day, we probably won't need crouton any more.

Agree, probably not using chroots but hopefully something similar in a container.

*Just my 2 cents,*
-**DennisL**
